### PR TITLE
feat: Add functions for lootrun beacon counts (and lootrun state)

### PR DESCRIPTION
**Pull request by kristofbolyai**
This is not a small PR. It adds a lot of stuff that can be used by us in the future:
- We can now track Wynn beacons
- We can track lootrun states
- We know how many of each beacon color we collected in the current lootrun

The only thing I did not do is persisting this in a storage, I will do that as a follow up.

### CODE REVIEW

File: common/src/main/java/com/wynntils/mc/mixin/ClientPacketListenerMixin.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/mixin/ClientPacketListenerMixin.java:568]: For those wondering (and so you dont have to open your IDEA while reviewing), this is the same mixin, but we are not calling it at RETURN, but at the last instruction. Original code is like

```
if (entity != null) ...
else log error
```

we need the entity local to still exist, and also this fixes null entities posting this event. win-win I guess.

File: common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java]: Obviously I got this from the resource pack, but I wanted to keep the format (damagePredicate * itemDurability)

**Comment by ryanzhoudev** [comment]
Lootrun ending from skill issue (dying too much, running out of time) does not reset the state

**Comment by kristofbolyai** [comment]
> Lootrun ending from skill issue (dying too much, running out of time) does not reset the state

That's weird.. Btw, Do you know if we should decrement the count if the user dies mid challenge?

**Comment by ryanzhoudev** [comment]
> > Lootrun ending from skill issue (dying too much, running out of time) does not reset the state
> 
> That's weird.. Btw, Do you know if we should decrement the count if the user dies mid challenge?

I am not sure if it counts towards the number of beacons selected by the user. I will try to find out

**Comment by ryanzhoudev** [comment]
We should not remove them, they still count towards total beacon numbers

File: common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java
Comment (ryanzhoudev) [inline common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java]: Technically this should be aqua

Comment (ryanzhoudev) [reply common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java]: Also we are missing green?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java]: Yes, you are right. I missed green because it is a different item, as it is also used for content tracking. Now this class is complete.

**Comment by kristofbolyai** [comment]
> Lootrun ending from skill issue (dying too much, running out of time) does not reset the state

I am not sure how you got this to happen. I checked timing out and it does reset. I assume dying does too, since we check for scoreboard removal.

File: common/src/main/java/com/wynntils/models/lootrun/LootrunModel.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/lootrun/LootrunModel.java:107]: I tried persisting but it is a bit harder than it looks.. So I will hold off on doing that for now. I will come back to it soon.

**Review by ryanzhoudev — APPROVED** [review]

**Review by kristofbolyai — COMMENTED** [review]
I self-reviewed this PR before merging. Obviously, it is not as good as a review from someone else, but DonkeyBlaster's and my review is enough for me to merge this. I am happy to do any post-merging cleanups, if needed.

File: common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java
Comment (magicus) [inline common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java:15]: Why these 0.xxx * 32 instead of normal integers?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java:15]: They are damage values taken from wynns resource pack

File: common/src/main/java/com/wynntils/models/beacons/BeaconModel.java
Comment (magicus) [inline common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:29]: Is this a good margin? How many are there typically? 7 or 27?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:29]: There are like 128/7.5 of them. This is more than enough.

File: common/src/main/java/com/wynntils/models/beacons/BeaconModel.java
Comment (magicus) [inline common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:110]: Maybe VectorUtils.equalsDiscardingY?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:110]: I guess yes, that can be added.

File: common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java
Comment (magicus) [reply common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java:15]: And there they are represented as floats times 32? 

Even if they are, can't we just replace them with the normal integer damage value? To me, a damage value of `3` makes sense, while `0.09375 * 32` does not.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java:15]: I wanted to preserve how the resource pack has it which is 0-1 damage * itemDurability. But if you want to, we can make these integers

File: common/src/main/java/com/wynntils/models/beacons/BeaconModel.java
Comment (magicus) [inline common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:62]: It could be that you are just a bit lucky that this works. Donkey introduced a tick delay in ShamanTotemModel:

```
                    // These must be ran with a delay, as inventory contents are set a couple ticks after the totem
                    // actually spawns
```

File: common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java
Comment (magicus) [inline common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java:76]: This actually means QUEST, right?

File: common/src/main/java/com/wynntils/models/beacons/BeaconModel.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:62]: I don't think so. They are always built up in order, and we also allow a delay. 1000 ms. See timed set.

File: common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/type/BeaconColor.java:76]: well caves do have beacons as well

**Pull request merged**
